sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+      the value of assert failure flag.
+   2. A host-task is enqueued to check value of assert failure flag.
+   3. The host task calls abort whenever assert failure flag is set.
+
+Illustrating this with an example, lets assume the user enqueues three kernels:
+ - `Kernel #1`
+ - `Kernel #2`
+ - `Kernel #3`, which depends on `Kernel #1`
+
+The resulting graph will look like this: ![graph](images/assert-fallback-graph.svg)
+
+### Interface to program scope variable
+
+Multiple translation units could be compiled/linked into a single device binary
+image. All of them should have `extern` declaration of program scope variable
+available. Definition of the variable is only available within devicelib in the
+same binary image where fallback `__devicelib_assert_fail` resides.
+
+<a name="prog-scope-var-decl">The variable has the following structure and
+declaration:</a>
+
+```c++
+struct AssertHappened {
+  int Flag = 0;
+};
+
+#ifdef __SYCL_DEVICE_ONLY__
+extern SYCL_GLOBAL_VAR AssertHappened AssertHappenedMem;
+#endif
+```
+
+Here, `SYCL_GLOBAL_VAR` is a macro which wraps special attribute to allow for
+mutable program-scope variable.
+
+The reference to extern variable is resolved within online-linking against
+fallback devicelib.
+
+### Online-linking fallback `__devicelib_assert_fail`
+
+Online linking against fallback implementation of `__devicelib_assert_fail` is
+performed only when assertion is enabled and Device-side Runtime doesn't provide
+implementation of `__devicelib_assert_fail`.
+
+In DPCPP headers one can see if assert is enabled with status of `NDEBUG` macro
+with `#ifdef`'s. This allows to enqueue a copy kernel and host task. The copy
+kernel will copy `AssertHappenedMem` to host and host-task will check the `Flag`
+value and `abort()` as needed.
+
+When in DPCPP Runtime Library this knowledge is obtained from device binary
+image descriptor's property sets.
+
+Each device image is supplied with an array of property sets:
+```c++
+struct pi_device_binary_struct {
+  //...
+  // Array of property sets
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+```
+Each property set is represented by the following struct:
+```c++
+// Named array of properties.
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+```
+It contains name of property set and array of properties. Each property is
+represented by the following struct:
+```c++
+struct _pi_device_binary_property_struct {
+  char *Name;       // null-terminated property name
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // _pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+```
+
+A distinct property set `SYCL/assert used` is added. In this set a property
```

Comment:
I'm okay to have this separate property set for now, but for a longer-term, I would like to unify this with handling of other SYCL 2020 aspects somehow.

This doesn't map 1:1 to existing aspects like `fp16`, because our intent is to provide the assert functionality regardless of device support for it. However, internally it might make sense to treat a call to assert as an aspect usage to unify the collection of aspects used in each kernel and minimize the amount of property sets.

Regardless of how it will be implemented, I think that it makes sense to add `native_assert`/`safe_assert` aspect to the SYCL extension specification, so users are able to check device capabilities and build their expectations from assert-based debugging accordingly.

Reply:
I plan to propose a property set named ""SYCL/kernel reqs", which hold requirements for each kernel.  My plan was to store each kernel's set of required aspects here, and also any requirements from `[[reqd_work_group_size()]]` and `[[reqd_sub_group_size()]]` attributes.  Possibly, we could use this property set also to tell whether a kernel needs support for assertions.

> Regardless of how it will be implemented, I think that it makes sense to add native_assert/safe_assert aspect to the SYCL extension specification, so users are able to check device capabilities and build their expectations from assert-based debugging accordingly.

I wonder if this would really be useful, though.  Are you thinking that applications may choose to disable assertions unless the device had native support?  Do we think that is likely?

I'm not opposed to adding the assertion, but I think we should only do so if we see a reasonable usage model for it.

Reply:
I didn't get the resolution here.
Is the following needed?

> Regardless of how it will be implemented, I think that it makes sense to add native_assert/safe_assert aspect to the SYCL extension specification, so users are able to check device capabilities and build their expectations from assert-based debugging accordingly.

Reply:
@gmlueck , @AlexeySachkov , a friendly ping on question above

Reply:
@AlexeySachkov, I think the question is to you.  If it were me, I think I'd wait until there is a user request for an aspect like `native_assert`.  However, I'm not opposed to adding it if you think there is a real use case.

Reply:
> I plan to propose a property set named ""SYCL/kernel reqs", which hold requirements for each kernel. My plan was to store each kernel's set of required aspects here, and also any requirements from [[reqd_work_group_size()]] and [[reqd_sub_group_size()]] attributes. Possibly, we could use this property set also to tell whether a kernel needs support for assertions.

Responding to my own comment.  I have now submitted a design document which adds the property set named "SYCL/requirements".  I think you could use this to say whether each kernel uses assertions.  See #3781.

Reply:
@gmlueck,

> I wonder if this would really be useful, though. Are you thinking that applications may choose to disable assertions unless the device had native support? Do we think that is likely?

I'm not sure that someone will disable assertions if native support is not available, but as a developer who added an assertion which didn't fire even if it should have been, I think I would appreciate some indicator somewhere which tells me: assertions are emulated and might not work properly in complex code. So I move on to other debugging mechanisms like printf/streams/writing to extra buffer/using actual debugger sooner rather than exploring my code once again trying to find out what went wrong at which stage so my `assert(false)` is not being displayed.

Reply:
I agree that there is a danger that users will be confused when the fall-back approach doesn't work and an assertion failure goes unreported.  However, I'm not sure that adding a new aspect will solve that problem.  I guess you're thinking that an application could test the aspect and then print a warning message about assertions being unreliable?  It seems unlikely to me that many people would do this, but who knows.

Here are some other possible solutions, but none look great to me:

* The runtime could print a warning message whenever a kernel that uses assertions is submitted to a device that doesn't support them natively.  However, I think users will complain about seeing the warning message all the time.

* The runtime could print a warning as above, but only if the kernel doesn't complete within X time.  This would reduce the frequency of the messages to only long-running kernels.

* We could create a new environment variable which needs to be set in order to print the warning.  However, I suspect that few people would think set the environment variable.

Reply:
What about having some tool which prints that info about the device? It could be `sycl-ls` (launched with some option), which we already ship it and we could extend some documentation saying that this tool can also be used to print various device capabilities to explore your platform in more details.

Reply:
> I guess you're thinking that an application could test the aspect and then print a warning message about assertions being unreliable?

As I developer, I would quickly hack my app to print that info to get a clue of what could go wrong - this is the use case I imagine. IDEs could somehow obtain this info as well through a tool to show some hints right next to assertions

Reply:
> What about having some tool which prints that info about the device? It could be sycl-ls [...]

This seems reasonable.  We would need an aspect to enable this, and that would also enable the other use cases you describe.  Now we just need to decide on a name for the aspect.  How about:

```
enum class aspect {
  // ...
  ext_intel_native_assert
};
```

If we decide to add this, it should go in the extension specification document (SYCL_ONEAPI_ASSERT.asciidoc).

Reply:
Added aspect to extension.

>Responding to my own comment. I have now submitted a design document which adds the property set named "SYCL/requirements". I think you could use this to say whether each kernel uses assertions. See #3781.

Will change to "SYCL/requirements" after #3781 merge.